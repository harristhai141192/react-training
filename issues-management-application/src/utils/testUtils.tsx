// Libraries
import React from 'react';
import { ThemeProvider } from '@chakra-ui/react';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';

// Themes
import themes from '@themes/index';

const customRender = (component: React.ReactNode) => {
  return render(
    <ThemeProvider theme={themes}>
      <Router>{component}</Router>
    </ThemeProvider>,
  );
};

export * from '@testing-library/react';
export { customRender as render };
