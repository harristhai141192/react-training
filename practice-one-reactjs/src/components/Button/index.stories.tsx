import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import Button from '.';

export default {
  title: 'Practice-One/Button',
  component: Button,
  argTypes: {
    backgroundColor: { control: 'color' },
  },
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  style: 'primary',
  label: 'primary',
};

export const Secondary = Template.bind({});
Secondary.args = {
  style: 'secondary',
  label: 'secondary',
};

export const Tertiary = Template.bind({});
Tertiary.args = {
  style: 'tertiary',
  label: 'tertiary',
};

export const Outlined = Template.bind({});
Outlined.args = {
  style: 'outlined',
  label: 'outlined',
};

export const Small = Template.bind({});
Small.args = {
  size: 'small',
  label: 'Button',
};

export const Medium = Template.bind({});
Medium.args = {
  size: 'medium',
  label: 'Button',
};

export const Large = Template.bind({});
Large.args = {
  size: 'large',
  label: 'Button',
};
