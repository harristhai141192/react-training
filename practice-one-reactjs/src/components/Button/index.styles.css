.defaultButton {
  font-family: DMSansMedium;
  border: 0;
  border-radius: 1.2em;
  cursor: pointer;
  display: inline-block;
}

.defaultButton:hover {
  color: white;
  background-color: black;
  transition: 0.4s;
  opacity: 0.6;
}

.primary {
  color: var(--default-button-font-color);
  background-color: var(--primary-button-background-color);
}

.secondary {
  color: var(--default-button-font-color);
  background-color: var(--secondary-button-background-color);
  box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px 1px inset;
}

.tertiary {
  color: var(--primary-button-font-color);
  background-color: var(--tertiary-button-background-color);
  box-shadow: rgba(0, 0, 0, 0.3) 0px 0px 0px 1px inset;
}

.outlined {
  color: var(--secondary-button-font-color);
  background-color: transparent;
  box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px 1px inset;
}

.small {
  font-size: var(--font-small);
  padding: 10px 16px;
}

.medium {
  font-size: var(--font-medium);
  padding: 11px 20px;
}

.large {
  font-size: var(--font-large);
  padding: 12px 24px;
}
