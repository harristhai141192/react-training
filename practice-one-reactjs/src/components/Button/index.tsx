import React from 'react';
import './index.styles.css';

interface IProps {
  type?: 'submit' | 'button' | 'reset';
  style?: 'primary' | 'secondary' | 'tertiary' | 'outlined';
  backgroundColor?: string;
  size?: 'small' | 'medium' | 'large';
  label: string;
  className?: string;
  children?: React.ReactNode;
  onClick?: () => void;
}

const Button = ({
  type,
  style = 'primary',
  size = 'medium',
  backgroundColor,
  label,
  className,
  onClick,
  ...props
}: IProps) => (
  <button
    type={type}
    className={['defaultButton', `${style}`, `${size}`, `${className}`].join(' ')}
    style={{ backgroundColor }}
    onClick={onClick}
    {...props}
  >
    {label}
  </button>
);

export default Button;
