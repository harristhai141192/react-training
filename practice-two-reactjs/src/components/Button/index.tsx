// Libraries
import React from 'react';
import { Button as ButtonChakra } from '@chakra-ui/react';
interface IProps {
  variant?: 'ghost' | 'outline' | 'solid' | 'link' | 'unstyled';
  colorScheme?: string;
  size?: 'xs' | 'sm' | 'md' | 'lg';
  label?: string;
  loadingText?: string;
  isLoading?: boolean;
  isDisabled?: boolean;
  className?: string;
  leftIcon?: React.ReactElement;
  name?: string;
  children?: React.ReactNode;
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
}

const Button = ({
  variant = 'solid',
  size = 'md',
  leftIcon,
  colorScheme,
  label,
  loadingText = '',
  isLoading,
  isDisabled,
  onClick,
  ...props
}: IProps) => (
  <ButtonChakra
    marginRight='5px'
    size={size}
    leftIcon={leftIcon}
    colorScheme={colorScheme}
    variant={variant}
    loadingText={loadingText}
    isDisabled={isDisabled}
    isLoading={isLoading}
    onClick={onClick}
    {...props}
  >
    {label}
  </ButtonChakra>
);

export default Button;
