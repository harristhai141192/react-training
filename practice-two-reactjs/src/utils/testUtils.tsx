import { ThemeProvider } from '@chakra-ui/react';
import { render } from '@testing-library/react';
import React from 'react';
import themes from '@themes/index';
import { BrowserRouter as Router } from 'react-router-dom';

const customRender = (component: React.ReactNode) => {
  return render(
    <ThemeProvider theme={themes}>
      <Router>{component}</Router>
    </ThemeProvider>,
  );
};

export * from '@testing-library/react';
export { customRender as render };
