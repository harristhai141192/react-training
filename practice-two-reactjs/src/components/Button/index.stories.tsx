// Libs
import { ComponentStory, ComponentMeta } from '@storybook/react';

// Component
import Button from '.';

import '../../index.css';

export default {
  title: 'Components/Button',
  component: Button,
  argTypes: {
    backgroundColor: { control: 'color' },
    size: {
      options: ['xs', 'sm', 'md', 'lg'],
      control: { type: 'select' },
    },
    variant: {
      options: ['ghost', 'outline', 'solid', 'link', 'unstyled'],
      control: { type: 'radio' },
    },
  },
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />;

export const Solid = Template.bind({});
Solid.args = {
  variant: 'solid',
  label: 'Try Now',
};

export const Primary = Template.bind({});
Primary.args = {
  variant: 'outline',
  label: 'Get A Demo',
};

export const Nav = Template.bind({});
Nav.args = {
  variant: 'ghost',
  label: 'Features',
};
