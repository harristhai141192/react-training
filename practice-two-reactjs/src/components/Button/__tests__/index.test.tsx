// libraries
import { render, fireEvent } from '@utils/testUtils';

// components
import Button from '../index';

const baseProps = {
  label: 'Button',
  isLoading: false,
};

const setup = (overrideProps = {}) => {
  const props = {
    ...baseProps,
    ...overrideProps,
  };
  return render(<Button {...props} />);
};

describe('Button component', () => {
  test('Component [Button] should match snapshot', () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  test('Component [Button] with on click function should be called once when click button', () => {
    const handleOnClick = jest.fn();
    const { getByText } = setup({ onClick: handleOnClick });
    fireEvent.click(getByText('Button'));
    expect(handleOnClick).toHaveBeenCalledTimes(1);
  });

  test('Component [Button] on click will be loading', () => {});
});
