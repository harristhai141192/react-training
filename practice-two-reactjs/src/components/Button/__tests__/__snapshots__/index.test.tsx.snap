// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Component [Button] should match snapshot 1`] = `
<div>
  <button
    class="chakra-button css-jcr1fc"
    type="button"
  />
</div>
`;
